File: puzzles/medium/cpp/last-crusade1.cpp
#include <iostream>
#include <vector>
#include <sstream>

using namespace std;

void printNextRoom(int roomType, int x, int y, string entrance) {
  if (roomType == 2 || roomType == 6) {
    if (entrance == "LEFT") {
      x++;
    } else {
      x--;
    }
  } else if (roomType == 4) {
    if (entrance == "TOP") {
      x--;
    } else {
      y++;
    }
  } else if (roomType == 5) {
    if (entrance == "TOP") {
      x++;
    } else {
      y++;
    }
  } else if (roomType == 10) {
    x--;
  } else if (roomType == 11) {
    x++;
  } else {
    y++;
  }
  
  cout << x << " " << y << endl;
}

int main() {
  vector<int> rooms;
  int nbColumns;
  int nbRows;
  cin >> nbColumns >> nbRows; cin.ignore();
  for (int i = 0; i < nbRows; i++) {
    string line;
    getline(cin, line);
    istringstream stream(line);
    int room;
    while (stream >> room) {
      rooms.push_back(room);
    }
  }
  int ex;
  cin >> ex; cin.ignore();

  // game loop
  while (true) {
    int x, y;
    string entrance;
    cin >> x >> y >> entrance; cin.ignore();
    printNextRoom(rooms[y * nbColumns + x], x, y, entrance);
  }
}

File: puzzles/ts/last-crusade1/last-crusade1.ts
function printNextRoom(roomType: number, x: number, y: number, entrance: string) {
  switch (roomType) {
    case 2: case 6:
      entrance === "LEFT" ? x++ : x--;
      break;
    case 4:
      entrance === "TOP" ? x-- : y++;
      break;
    case 5:
      entrance === "TOP" ? x++ : y++;
      break;
    case 10:
      x--;
      break;
    case 11:
      x++;
      break;
    default:
      y++;
  }
  console.log(`${x} ${y}`);
}

let rooms: number[] = [];
let inputs: string[] = readline().split(' ');
const nbColumns: number = Number(inputs[0]);
const nbRows: number = Number(inputs[1]);
for (let i: number = 0; i < nbRows; i++) {
  let line = readline().split(' ').map(Number);
  rooms.push(line);
}
readline();

// game loop
while (true) {
  const inputs: string[] = readline().split(' ');
  const x: number = Number(inputs[0]);
  const y: number = Number(inputs[1]);
  const entrance: string = inputs[2];
  printNextRoom(rooms[y][x], x, y, entrance);
}
